hazelcast/src/main/java/com/hazelcast/json/internal/JsonSchemaHelper.java

```diff
@@ -0,0 +1,229 @@
+            JsonSchemaStructNode structDescription = (JsonSchemaStructNode) schemaNode;
+            if (structDescription.getChildCount() <= expectedOrderIndex) {
+                return null;
+            }
+            JsonSchemaNameValue nameValue = structDescription.getChild(expectedOrderIndex);
+            if (structMatches(input, nameValue, expectedOrderIndex, attributePath)) {
+                schemaNode = nameValue.getValue();
+            } else {
+                return null;
+            }
+        }
+        if (schemaNode.isTerminal() && attributePath.getNext() == null) {
+            // at this point we are sure we found the value by pattern. So we have to be able to extract JsonValue.
+            // Otherwise, let the exceptions propagate
+            try {
+                JsonReducedValueParser valueParser = new JsonReducedValueParser();
+                int valuePos = ((JsonSchemaTerminalNode) schemaNode).getValueStartLocation();
+                return input.parseValue(valueParser, valuePos);
+            } catch (ParseException parseException) {
+                throw new HazelcastException(parseException);
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Creates a description out of a JsonValue. The parser must be
+     * pointing to the start of the input.
+     *
+     * @param parser
+     * @return
+     * @throws IOException
+     */
+    public static JsonSchemaNode createSchema(JsonParser parser) throws IOException {
+        JsonSchemaNode dummy = new JsonSchemaStructNode(null);
+        JsonSchemaStructNode parent = (JsonSchemaStructNode) dummy;
+        JsonToken currentToken = parser.nextToken();
+        int nameLocation = -1;
+        while (currentToken != null) {
+            if (currentToken.isStructStart()) {
+                JsonSchemaStructNode structNode = new JsonSchemaStructNode(parent);
+                JsonSchemaNameValue nameValue = new JsonSchemaNameValue(nameLocation, structNode);
+                parent.addChild(nameValue);
+                parent = structNode;
+                nameLocation = -1;
+            } else if (currentToken == JsonToken.FIELD_NAME) {
+                nameLocation = (int) getLocation(parser.getTokenLocation());
+            } else if (currentToken.isStructEnd()) {
+                parent = parent.getParent();
+                nameLocation = -1;
+            } else {
+                JsonSchemaTerminalNode terminalNode = new JsonSchemaTerminalNode(parent);
+                terminalNode.setValueStartLocation((int) getLocation(parser.getTokenLocation()));
+                JsonSchemaNameValue nameValue = new JsonSchemaNameValue(nameLocation, terminalNode);
+                parent.addChild(nameValue);
+                nameLocation = -1;
+            }
+            currentToken = parser.nextToken();
+        }
+        JsonSchemaNameValue nameValue = ((JsonSchemaStructNode) dummy).getChild(0);
```

Comment:
Minor: in an extreme case when the provided JSON is an empty string, this will fail with `IndexOutOfBoundsException`.

Reply:
added a check and JsonSchemaHelper#testEmptyStringReturnsNullSchema test.